token/program-2022/src/extension/cpi_guard/instruction.rs

```diff
@@ -0,0 +1,100 @@
+use {
+    crate::{
+        check_program_account,
+        instruction::{encode_instruction, TokenInstruction},
+    },
+    num_enum::{IntoPrimitive, TryFromPrimitive},
+    solana_program::{
+        instruction::{AccountMeta, Instruction},
+        program_error::ProgramError,
+        pubkey::Pubkey,
+    },
+};
+
+/// CPI Guard extension instructions
+#[derive(Clone, Copy, Debug, PartialEq, IntoPrimitive, TryFromPrimitive)]
+#[repr(u8)]
+pub enum CpiGuardInstruction {
+    /// Lock certain token operations from taking place within CPI for this Account, namely:
+    /// * Transfer and Burn must go through a delegate.
+    /// * CloseAccount can only return lamports to owner.
+    /// * SetAuthority can only be used to remove an existing close authority.
+    /// * Approve is disallowed entirely.
+    ///
+    /// In addition, CPI Guard cannot be enabled or disabled via CPI.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[writable]` The account to update.
+    ///   1. `[signer]` The account's owner.
+    ///
+    ///   * Multisignature authority
+    ///   0. `[writable]` The account to update.
+    ///   1. `[]` The account's multisignature owner.
+    ///   2. ..2+M `[signer]` M signer accounts.
+    ///
+    Enable,
+    /// Allow all token operations to happen via CPI as normal.
+    ///
+    /// Fails if the account does not have the extension present.
```

Comment:
The implementation does not follow this comment, since `disable_cpi_guard` also initializes the extension.  I don't have a strong opinion either way, but I think it's nicer to initialize it even in the case of `disable_cpi_guard`, which is what memo transfer does too.  We'll have to update that other comment too